sdk/keyvault/Azure.Security.KeyVault.Keys/src/Base64Url.cs

```diff
@@ -0,0 +1,25 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for
+// license information.
+
+using System;
+using System.Text;
+
+namespace Azure.Security.KeyVault.Keys
+{
+    internal static class Base64Url
+    {
+        public static byte[] Decode(string str)
+        {
+            str = new StringBuilder(str).Replace('-', '+').Replace('_', '/').Append('=', (str.Length % 4 == 0) ? 0 : 4 - (str.Length % 4)).ToString();
+
+            return Convert.FromBase64String(str);
+        }
+
+        public static string Encode(byte[] bytes)
+        {
+            return new StringBuilder(Convert.ToBase64String(bytes)).Replace('+', '-').Replace('/', '_').Replace("=", "").ToString();
```

Comment:
If performance is important here, I believe this can be optimized quite a bit, especially if we write UTF-8 data directly.

One step could be to use the `System.Buffers.Text.Base64` APIs to write `Base64` as UTF-8 text and then modify to be `Base64Url` encoded. A better approach could be to use `Base64Url` encoding directly to avoid the string allocations and double encoding. This one requires a bit more work since we don't have one that's built-in to .NET. There's [WebEncoders](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.webutilities.webencoders.base64urlencode?view=aspnetcore-2.2 ) utility within aspnet.

Reply:
To clarify, something like this could make a huge impact in terms of runtime perf and allocations (at least as a stop gap). And this is not accounting for the fact that you save some of the cost of writing UTF-16 string as JSON (since it's already UTF-8).

A built-in API can be further optimized (with vectorization, and direct Base64 URL encoding rather than going through Base64), but this would bridge the gap quite a bit.

**Note:** The implementation is just an example and not tested (so you probably want to do some due diligence).
``` ini

BenchmarkDotNet=v0.11.3.1003-nightly, OS=Windows 10.0.18362
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-preview4-010907
  [Host]     : .NET Core 3.0.0-preview6-27718-08 (CoreCLR 3.0.19.26872, CoreFX 4.700.19.26805), 64bit RyuJIT
  DefaultJob : .NET Core 3.0.0-preview4-27519-01 (CoreCLR 4.6.27519.71, CoreFX 4.7.19.16310), 64bit RyuJIT


```
|        Method | Length |        Mean |     Error |    StdDev |      Median |         Min |         Max | Ratio | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|-------------- |------- |------------:|----------:|----------:|------------:|------------:|------------:|------:|------------:|------------:|------------:|--------------------:|
| **CurrentEncode** |     **20** |   **293.36 ns** |  **6.207 ns** |  **5.502 ns** |   **292.35 ns** |   **288.06 ns** |   **309.31 ns** |  **1.00** |      **0.0801** |           **-** |           **-** |               **336 B** |
|     NewEncode |     20 |    65.21 ns |  1.371 ns |  1.215 ns |    65.02 ns |    62.66 ns |    67.48 ns |  0.22 |      0.0134 |           - |           - |                56 B |
|               |        |             |           |           |             |             |             |       |             |             |             |                     |
| **CurrentEncode** |    **200** | **1,904.35 ns** | **39.632 ns** | **37.072 ns** | **1,899.68 ns** | **1,850.80 ns** | **1,978.77 ns** |  **1.00** |      **0.4234** |           **-** |           **-** |              **1776 B** |
|     NewEncode |    200 |   477.29 ns |  6.333 ns |  5.614 ns |   476.49 ns |   466.32 ns |   488.64 ns |  0.25 |      0.0706 |           - |           - |               296 B |


```C#
private static string Encode(byte[] bytes)
{
    return new StringBuilder(Convert.ToBase64String(bytes)).Replace('+', '-').Replace('/', '_').Replace("=", "").ToString();
}

private static ReadOnlySpan<byte> EncodeToUtf8Bytes(byte[] bytes)
{
    byte[] utf8Text = new byte[Base64.GetMaxEncodedToUtf8Length(bytes.Length)];
    Base64.EncodeToUtf8(bytes, utf8Text, out int bytesConsumed, out int bytesWritten);
    Debug.Assert(bytesConsumed == bytes.Length);

    if (utf8Text[bytesWritten - 1] == (byte)'=')
    {
        bytesWritten--;
        if (utf8Text[bytesWritten - 1] == (byte)'=')
        {
            bytesWritten--;
        }
    }

    Debug.Assert(utf8Text[bytesWritten - 1] != (byte)'=');

    for (int i = 0; i < bytesWritten; i++)
    {
        byte val = utf8Text[i];
        if (val <= (byte)'/')
        {
            if (val == '+')
            {
                utf8Text[i] = (byte)'-';
            }
            else
            {
                Debug.Assert(val == '/');
                utf8Text[i] = (byte)'_';
            }
        }
    }

    return utf8Text.AsSpan(0, bytesWritten);
}
```

Reply:
This is great information. Thank you! I created issue https://github.com/Azure/azure-sdk-for-net/issues/6476 so we can follow on how to improve the current implementation